tests/test_app_bokeh.py

```diff
@@ -0,0 +1,741 @@
+        plt.show()
+    return np.array(Image.open(im))
+
+
+def get_renderer_prop(prop: str) -> json:
+    """Get a renderer property from the server."""
+    resp = main.UI["s"].get(f"http://{main.host2}:5000/tileserver/renderer/{prop}")
+    return resp.json()
+
+
+@pytest.fixture(scope="module", autouse=True)
+def annotation_path(data_path: dict[str, object]) -> dict[str, object]:
+    """Download some testing slides and overlays.
+
+    Sets up a dictionary defining the paths to the files
+    that can be grabbed as a fixture to refer to during tests.
+
+    """
+    data_path["slide1"] = _fetch_remote_sample(
+        "svs-1-small",
+        data_path["base_path"] / "slides",
+    )
+    data_path["slide2"] = _fetch_remote_sample(
+        "ndpi-1",
+        data_path["base_path"] / "slides",
+    )
+    data_path["slide3"] = _fetch_remote_sample(
+        "patch-extraction-vf",
+        data_path["base_path"] / "slides",
+    )
+    data_path["annotations"] = _fetch_remote_sample(
+        "annotation_store_svs_1",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["graph"] = _fetch_remote_sample(
+        "graph_svs_1",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["graph_feats"] = _fetch_remote_sample(
+        "graph_svs_1_feats",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["img_overlay"] = _fetch_remote_sample(
+        "rendered_annotations_svs_1",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["geojson_anns"] = _fetch_remote_sample(
+        "geojson_cmu_1",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["dat_anns"] = _fetch_remote_sample(
+        "annotation_dat_svs_1",
+        data_path["base_path"] / "overlays",
+    )
+    data_path["config"] = _fetch_remote_sample(
+        "config_2",
+        data_path["base_path"] / "overlays",
+    )
+    return data_path
+
+
+def run_app() -> None:
+    """Helper function to launch a tileserver."""
+    app = TileServer(
+        title="Tiatoolbox TileServer",
+        layers={},
+    )
+    CORS(app, send_wildcard=True)
+    app.run(host="127.0.0.1", threaded=False)
+
+
+@pytest.fixture(scope="module")
+def doc(data_path: dict[str, object]) -> Generator[Document, object, None]:
+    """Create a test document for the visualization tool."""
+    # start tile server
+    p = multiprocessing.Process(target=run_app, daemon=True)
+    p.start()
+    time.sleep(2)  # allow time for server to start
+
+    main.doc_config.set_sys_args(argv=["dummy_str", str(data_path["base_path"])])
+    handler = FunctionHandler(main.doc_config.setup_doc)
+    app = Application(handler)
+    yield app.create_document()
+    p.terminate()
+
+
+# test some utility functions
+
+
+def test_to_num() -> None:
+    """Test the to_num function."""
+    assert main.to_num("1") == 1
+    assert main.to_num("1.0") == 1.0
+    assert main.to_num("1.0e-3") == 1.0e-3
+    assert main.to_num(2) == 2
```

Comment:
```suggestion
    assert main.to_num("2") == 2
```

Reply:
this is intentional, to make sure if it is passed a number it will return it unchanged

Reply:
I've changed the type annotations for this function to `str | SupportsFloat`